### PR TITLE
remove url encoding from url request

### PR DESCRIPTION
**Pull request by vhvb1989**
Adding a new internal API to calculate the size of one span after it gets url-encoded.   
This method is used by our curl transport adapter to know how many bytes to allocate for a url.

*Adding internal header into curl transport adapter to consume az_span_url_encode(...)..  
I think this is OK since we are the ones providing the transport adapter, so we can access internal stuff.

Some one else implementing a transport adapter would need to make sure to url-enconde the url from http request somehow. (or we could expose _az_span_url_enconde to public API in the future if we want/customer ask for it)

fixes: https://github.com/Azure/azure-sdk-for-c/issues/963

### CODE REVIEW

File: sdk/inc/azure/core/internal/az_span_internal.h [hunk 1]
```diff
@@ -45,6 +45,17 @@ AZ_INLINE AZ_NODISCARD int32_t _az_span_diff(az_span sliced_span, az_span origin
 AZ_NODISCARD az_result
 _az_span_url_encode(az_span destination, az_span source, int32_t* out_length);
 
+/**
+ * @brief Calculates what would it be the length of \p source #az_span after url-encoding it.
+ *
+ * @param[in] source The #az_span containing the non-URL-encoded bytes.
+ * @param[out] out_length A pointer to an int32_t that is going to be assigned the length
+ * of URL-encoding the \p source.
```
Comment (ahsonkhan) [inline sdk/inc/azure/core/internal/az_span_internal.h:53]: Remove. This is probably what CI is complaining about.
```suggestion
```

File: sdk/inc/azure/core/internal/az_span_internal.h [hunk 2]
```diff
@@ -45,6 +45,17 @@ AZ_INLINE AZ_NODISCARD int32_t _az_span_diff(az_span sliced_span, az_span origin
...
+ * @return the length of source if it would be url-encoded.
```
Comment (ahsonkhan) [inline sdk/inc/azure/core/internal/az_span_internal.h:54]: nit:
```suggestion
 * @return The length of source if it would be url-encoded.
```

File: sdk/src/azure/core/az_aad.c [hunk 3]
```diff
@@ -49,18 +46,18 @@ AZ_NODISCARD az_result _az_aad_build_body(
   AZ_RETURN_IF_NOT_ENOUGH_SIZE(body, az_span_size(grant_type_and_client_id_key));
 
   az_span remainder = az_span_copy(body, grant_type_and_client_id_key);
-  int32_t url_length = 0;
+  int32_t body_length = 0;
 
-  AZ_RETURN_IF_FAILED(_az_span_url_encode(remainder, client_id, &url_length));
-  remainder = az_span_slice_to_end(remainder, url_length);
+  AZ_RETURN_IF_FAILED(_az_span_url_encode(remainder, client_id, &body_length));
```
Comment (ahsonkhan) [inline sdk/src/azure/core/az_aad.c:51]: @antkmsft, we should add a comment explaining why we are url encoding the body.

File: sdk/src/azure/core/az_span.c [hunk 4]
```diff
@@ -983,6 +983,34 @@ AZ_NODISCARD AZ_INLINE bool _az_span_url_should_encode(uint8_t c)
...
+{
+  _az_PRECONDITION_VALID_SPAN(source, 0, true);
+
+  int32_t const source_size = az_span_size(source);
+  if (source_size == 0)
+  {
+    return 0;
+  }
+
+  // We may or may not have enough space, given whether the input needs much encoding or not.
```
Comment (ahsonkhan) [inline sdk/src/azure/core/az_span.c:996]: There is no destination. What does it mean by enough space, in this context? I would reword this comment or remove it.

File: sdk/src/azure/core/az_span.c [hunk 5]
```diff
@@ -983,6 +983,34 @@ AZ_NODISCARD AZ_INLINE bool _az_span_url_should_encode(uint8_t c)
...
+  uint8_t* const src_ptr = az_span_ptr(source);
+  int32_t extra_space_used = 0;
```
Comment (ahsonkhan) [inline sdk/src/azure/core/az_span.c:998]: No space is being used here. Change local variable name.

File: sdk/src/azure/platform/az_curl.c [hunk 6]
```diff
@@ -480,7 +465,7 @@ _az_http_client_curl_setup_url(CURL* ref_curl, _az_http_request const* request)
   az_span request_url = { 0 };
   // get request_url. It will have the size of what it has written in it only
   AZ_RETURN_IF_FAILED(az_http_request_get_url(request, &request_url));
-  int32_t request_url_size = az_span_size(request_url);
+  int32_t request_url_size = _az_span_url_encode_calc_length(request_url);
```
Comment (ahsonkhan) [inline sdk/src/azure/platform/az_curl.c:468]: It's somewhat unfortunate that we have to do a double pass over the URL, once to calculate the length and again to actually encode it.

I wonder if it is worth exploring some alternatives:
1) Always assume worst case and malloc 3x unescaped url size.
2) Have the `_az_span_url_encode_calc_length` return the index when the first character to be url encoded is found, so at least we save the double processing until then.
3) Maybe we use a temporary scratch buffer and encode into that up-front, prior to the call to malloc.

File: sdk/tests/core/test_az_url_encode.c [hunk 7]
```diff
@@ -84,6 +85,10 @@ static void test_url_encode_basic(void** state)
     assert_true(az_span_is_content_equal(
         AZ_SPAN_FROM_BUFFER(buf20), AZ_SPAN_FROM_STR("aBc%2Fg*************")));
   }
+  {
+    int32_t url_length = _az_span_url_encode_calc_length(AZ_SPAN_FROM_STR("aBc/g"));
+    assert_int_equal(url_length, sizeof("aBc%2Fg") - 1);
```
Comment (ahsonkhan) [inline sdk/tests/core/test_az_url_encode.c:90]: Add more test cases? I would consider adding a call to `_az_span_url_encode_calc_length` and asserting the expected length for all the test strings we call `_az_span_url_encode` to in the existing tests.

Cases such as:
- empty span
- nothing is encoded
- first/last character is encoded
- every character is encoded
- % character, which in itself is encoded as %25

File: sdk/src/azure/platform/az_curl.c (hunk 6 above)
Comment (vhvb1989) [reply sdk/src/azure/platform/az_curl.c:468]: Well, we are targeting small devices,  I would rather same memory than making a double pass.

File: sdk/inc/azure/core/internal/az_span_internal.h [hunk 8]
```diff
@@ -45,6 +45,15 @@ AZ_INLINE AZ_NODISCARD int32_t _az_span_diff(az_span sliced_span, az_span origin
 AZ_NODISCARD az_result
 _az_span_url_encode(az_span destination, az_span source, int32_t* out_length);
 
+/**
+ * @brief Calculates what would it be the length of \p source #az_span after url-encoding it.
```
Comment (ahsonkhan) [inline sdk/inc/azure/core/internal/az_span_internal.h:49]:
```suggestion
 * @brief Calculates what would be the length of \p source #az_span after url-encoding it.
```

File: sdk/src/azure/core/az_span.c [hunk 9]
```diff
@@ -983,6 +983,33 @@ AZ_NODISCARD AZ_INLINE bool _az_span_url_should_encode(uint8_t c)
...
+  uint8_t* const src_ptr = az_span_ptr(source);
+  int32_t required_symbols_to_be_added = 0;
+  int32_t src_idx = 0;
+  do
+  {
+    uint8_t c = src_ptr[src_idx];
+    if (_az_span_url_should_encode(c))
+    {
+      required_symbols_to_be_added
+          += 2; // Adding '%' plus 2 digits (minus 1 as original symbol is counted as 1)
```
Comment (ahsonkhan) [inline sdk/src/azure/core/az_span.c:1005]: nit: Move the comment above the line so the code fits on one line.
```suggestion
      // Adding '%' plus 2 digits (minus 1 as original symbol is counted as 1)
      required_symbols_to_be_added += 2; 
```

File: sdk/tests/core/test_az_url_encode.c [hunk 10]
```diff
@@ -84,6 +85,42 @@ static void test_url_encode_basic(void** state)
...
+  }
+  {
+    // Empty span
+    int32_t url_length = _az_span_url_encode_calc_length(AZ_SPAN_NULL);
```
Comment (ahsonkhan) [inline sdk/tests/core/test_az_url_encode.c:94]: I was thinking of adding these next to each of the `_az_span_url_encode` calls in the tests, like here:
https://github.com/Azure/azure-sdk-for-c/blob/ce4d71bbd0f8a49ab79a47a5f23222bb659773c6/sdk/tests/core/test_az_url_encode.c#L33
https://github.com/Azure/azure-sdk-for-c/blob/ce4d71bbd0f8a49ab79a47a5f23222bb659773c6/sdk/tests/core/test_az_url_encode.c#L82
https://github.com/Azure/azure-sdk-for-c/blob/ce4d71bbd0f8a49ab79a47a5f23222bb659773c6/sdk/tests/core/test_az_url_encode.c#L152


But this is fine too.

File: sdk/tests/core/test_az_url_encode.c [hunk 11]
```diff
@@ -84,6 +85,42 @@ static void test_url_encode_basic(void** state)
...
+  {
+    // % character, which in itself is encoded as %25
+    int32_t url_length = _az_span_url_encode_calc_length(AZ_SPAN_FROM_STR("%"));
+    assert_int_equal(url_length, 3);
+  }
+  {
+    // a single length span with the \0 NULL character
+    int32_t url_length = _az_span_url_encode_calc_length(AZ_SPAN_FROM_STR("\0"));
+    assert_int_equal(url_length, 3);
+  }
```
Comment (ahsonkhan) [inline sdk/tests/core/test_az_url_encode.c:123]: nit: Move all these tests as a group to the end of the method, after this line:
https://github.com/Azure/azure-sdk-for-c/blob/ce4d71bbd0f8a49ab79a47a5f23222bb659773c6/sdk/tests/core/test_az_url_encode.c#L240-L243

Right now they are right in the middle between a bunch of `_az_span_url_encode` tests

File: sdk/src/azure/core/az_span.c [hunk 12]
```diff
@@ -983,6 +983,33 @@ AZ_NODISCARD AZ_INLINE bool _az_span_url_should_encode(uint8_t c)
...
+    uint8_t c = src_ptr[src_idx];
+    if (_az_span_url_should_encode(c))
+    {
+      // Adding '%' plus 2 digits (minus 1 as original symbol is counted as 1)
+      required_symbols_to_be_added += 2; 
+    }
+    ++src_idx;
+  } while (src_idx < source_size);
+
+  return source_size + required_symbols_to_be_added;
```
Comment (ahsonkhan) [inline sdk/src/azure/core/az_span.c:1010]: I just realized this and ` required_symbols_to_be_added += 2;` have potential integer overflow problem for really large inputs.

**Some ways to solve this:**
1) Add a precondition that the input must be of size < int.max/3 - I think this is best given the circumstance (add a comment too)
2) Add integer overflow checks and fail with some `az_result`
3) Add integer overflow checks and fail with -1

Comment (vhvb1989) [reply sdk/src/azure/core/az_span.c:1010]: This is one of the things we should not pay too much attention, IMO.
int32 max is 2,147,483,647.   divided by 3 is `715,827,882`.   
We would need an url length bigger than that value to hit this edge case.

Just think about it.  If we consider the average word count in a book is 700, 000 words, and suppose an average of 10 chars per word, we would have an entire book with 7, 000, 000 chars right?  Well, that would be only the 1% of the MAX supported url.

Comment (vhvb1989) [reply sdk/src/azure/core/az_span.c:1010]: https://stackoverflow.com/questions/417142/what-is-the-maximum-length-of-a-url-in-different-browsers

Comment (ahsonkhan) [reply sdk/src/azure/core/az_span.c:1010]: I agree with your sentiment.

That said we should be resilient to developer bugs, and edge cases with untrusted input. Adding a precondition conveys intent clearly and helps catch bugs where the user accidentally passes in corrupted or too large span (let's say they made a mistake with ptrs and size calculations (or overflowed it), even if the buffer itself isn't 800MB).

This is a very cheap check and I see no harm in adding it.

What is the behavior of this function when size is > 800 MB? Right now, undefined and error prone.
Add a precondition: Now the caller knows that size > 800 MB isn't supported.

Comment (vhvb1989) [reply sdk/src/azure/core/az_span.c:1010]: precondition is cheap.
Testcase is not.  I would need to allocate all that memory.  It is not required

Comment (ahsonkhan) [reply sdk/src/azure/core/az_span.c:1010]: I would be OK with not having a test case for it. Generally, we have a lower test coverage bar when it comes to preconditions, since they themselves are a sort of check. Covering those is nice to have.

One way you can make the test (if you really wanted to), is create a busted az_span with a "fake" large size and pass that in.

Comment (RickWinter) [reply sdk/src/azure/core/az_span.c:1010]: Adding a precondition would be good

File: sdk/src/azure/core/az_span.c [hunk 13]
```diff
@@ -983,6 +983,36 @@ AZ_NODISCARD AZ_INLINE bool _az_span_url_should_encode(uint8_t c)
   }
 }
 
+AZ_NODISCARD int32_t _az_span_url_encode_calc_length(az_span source)
+{
+  _az_PRECONDITION_VALID_SPAN(source, 0, true);
+  // trying to calculate the number of bytes to encode more than INT32_MAX / 3 might overflow an
+  // int32 and return an erroneous number back
+  _az_PRECONDITION_RANGE(0, az_span_size(source), INT32_MAX / 3);
```
Comment (ahsonkhan) [inline sdk/src/azure/core/az_span.c:991]: Great comment :)

File: sdk/tests/core/test_az_url_encode.c [hunk 14]
```diff
@@ -337,6 +374,16 @@ static void test_url_encode_preconditions(void** state)
...
+    {
+      // precondition -> bigger than INT32_MAX / 3
+      az_span simulate_span = { ._internal = { .ptr = NULL, .size = INT32_MAX / 3 + 1 } };
+      ASSERT_PRECONDITION_CHECKED(_az_span_url_encode_calc_length(simulate_span));
+    }
+    {
+      // precondition -> less than 0
+      az_span simulate_span = { ._internal = { .ptr = NULL, .size = -1 } };
+      ASSERT_PRECONDITION_CHECKED(_az_span_url_encode_calc_length(simulate_span));
+    }
```
Comment (ahsonkhan) [inline sdk/tests/core/test_az_url_encode.c:386]: Bonus points for the tests.

File: sdk/src/azure/core/az_aad.c [hunk 15]
```diff
@@ -21,11 +21,8 @@ _az_aad_build_url(az_span url, az_span authority, az_span tenant_id, az_span* ou
   AZ_RETURN_IF_NOT_ENOUGH_SIZE(url, az_span_size(authority));
   az_span remainder = az_span_copy(url, authority);
 
-  {
-    int32_t url_length = 0;
-    AZ_RETURN_IF_FAILED(_az_span_url_encode(remainder, tenant_id, &url_length));
-    remainder = az_span_slice_to_end(remainder, url_length);
```
Comment (ahsonkhan) [inline sdk/src/azure/core/az_aad.c:27]: With the changes from https://github.com/Azure/azure-sdk-for-c/pull/1022, we are no longer url encoding `tenant_id`. @vhvb1989 - can you fix this by reverting this part of the change, or is that not necessary?
